### PR TITLE
Run benchmarks in CI, resolve hypothesis errors

### PR DESCRIPTION
**Pull request by kdk**
Test PR to look at the feasibility of running asv benchmarks for PR branches in travis. Looking to find out if the variability of other tenants on the host hardware of the travis VM will wash out any meaningful information in the benchmark comparison.

### CODE REVIEW

**Comment by kdk** [comment]
This is ready for review. 
- Adds a travis job to run asv benchmarks comparing each PR branch to master (and master to last-master on a merge comit.) Excludes the `quantum_volume` and `random_circuit_hex` benchmarks to make sure we finish under the 50m job limit. There's a lot of variability in the performance on travis so these numbers will not be comparable outside a single job. Example output: https://travis-ci.com/Qiskit/qiskit-terra/jobs/209372043#L1064
- Renames the python 3.6 linux job to add the fact that it runs `coverage` and `diff-cover`.
- Returns the printing of QASM during transpiler equivalence testing. For now, this will give a quicker indication of how the tests are progressing than 20 minutes of silence.
- Removes the `hypothesis.errors.FailedHealthCheck` error check that occurs frequently on master (see most of the errors in #2645 , https://hypothesis.readthedocs.io/en/latest/healthchecks.html#hypothesis.HealthCheck.filter_too_much ). It's not clear if this warning is expected with hypothesis state-based testing, or if something in our `assume`/`precondition`/... usage is causing our tests to filter out more data. Data generation won't be the slow step in these tests anyway so it should be okay to ignore.

**Comment by ajavadia** [comment]
This looks good to me. @mtreinish can you check the benchmark runs?

File: .travis.yml [hunk 1]
```diff
@@ -196,9 +196,29 @@ jobs:
...
+      <<: *stage_linux
+      script:
+      - python setup.py build_ext --inplace
+      - pip install "qiskit-aer"
+      - pip install "asv"
+      - git clone "https://www.github.com/Qiskit/qiskit" meta-qiskit
+      - >-
+        sed -i
+        -e 's/"repo": "https:\/\/github.com\/Qiskit\/qiskit-terra.git",/"repo": "..",/'
+        -e 's/"pythons": \["3.5", "3.6", "3.7"\],/"pythons": ["3.5"],/'
```
Comment (mtreinish) [inline .travis.yml]: We can just specify this on the asv command with `-E pythons:3.5` (or something like that) instead of editing the config file.

**Comment by ajavadia** [comment]
Just so that we catch significant performance regressions before a merge, rather than after.

**Comment by mtreinish** [comment]
> Just so that we catch significant performance regressions before a merge, rather than after.

Sure, that's a good goal to have, but running it in travis or any CI backed by a public cloud there is no actual way to know. The data is super noisy. You don't actually know if any regression is caused by a patch or you just happened to get a noisy neighbor bitcoin mining. Or the vm is on a host with a failing disk or some other fault. Even extreme outliers will be hard to attribute to a single (or small number) run in travis.

**Comment by kdk** [comment]
The hope is that this will be useful as a code review tool, for now, since our confidence in the statistics won't be great (thus `allow_failures`) and to get some more exposure to the benchmarks (since they're not trivial to run locally.) 

Ideally, it would be great to address both of these, by getting dedicated hardware to run PR benchmarks/randomize tests/..., and by making the benchmarks easier to run from within terra. Running both in travis was a short term answer given our current extra travis capacity.

**Comment by mtreinish** [comment]
If we wanted to do pre-merge performance testing for real, that is something we can do, it's just a lot more work. We'd have to brush the dust off of: https://github.com/mtreinish/asv-github and rework it with the assumption of being triggered by PRs. Then I can deploy that on the dedicated benchmark machine we have and have it use whatever idle time it has to run benchmarks for proposed PRs and report back the results. But that's a bigger chunk of work, and IMO is a lower priority than other tasks.

**Comment by kdk** [comment]
Agree that's the right direction to head, and also substantially more work. As is, would asv in travis be useful for e.g. evaluating the performance changes of PRs like #2529 ?

**Comment by ajavadia** [comment]
Yes I agree it can be noisy. However the noise *might* be mitigated or averaged out since @kdk is running 
```
benchmarkA pre  --> benchmarkA post --> benchmarkB pre --> benchmarkB post --> ...
                 x                                     x
```
so we will be screwed if bitcoin mining happens at the `x` marks, but I was hoping the chances of that are low.

**Pull request merged**